File: solutions/swift/twelve-days/7/Sources/TwelveDays/TwelveDays.swift
class TwelveDaysSong {
  private static let ordinals = [
    "first", "second", "third", "fourth", "fifth", "sixth", "seventh", "eighth", "ninth", "tenth",
    "eleventh", "twelfth",
  ]
  private static let subjects = [
    "a Partridge", "two Turtle Doves", "three French Hens", "four Calling Birds", "five Gold Rings",
    "six Geese-a-Laying", "seven Swans-a-Swimming", "eight Maids-a-Milking", "nine Ladies Dancing",
    "ten Lords-a-Leaping", "eleven Pipers Piping", "twelve Drummers Drumming",
  ]
  private static func joinSubjects(_ n: Int) -> String {
    if n == 0 { return subjects[0] }
    if n == 1 { return subjects[1] + ", and " + subjects[0] }
    return subjects[n] + ", " + joinSubjects(n - 1)
  }
  private static func verse(_ v: Int) -> String {
    "On the \(ordinals[v-1]) day of Christmas my true love gave to me: \(joinSubjects(v-1)) in a Pear Tree."
  }
  static func recite(start: Int, end: Int) -> String {
    (start...end).map(verse).joined(separator: "\n")
  }
}

File: solutions/lua/twelve-days/2/twelve-days.lua
local ordinals = { "first", "second", "third", "fourth", "fifth", "sixth", "seventh", "eighth", "ninth", "tenth", "eleventh", "twelfth" }
local subjects = { "twelve Drummers Drumming, ", "eleven Pipers Piping, ", "ten Lords-a-Leaping, ", "nine Ladies Dancing, ", "eight Maids-a-Milking, ", "seven Swans-a-Swimming, ", "six Geese-a-Laying, ", "five Gold Rings, ", "four Calling Birds, ", "three French Hens, ", "two Turtle Doves, and ", "a Partridge" }
local verse = "On the %s day of Christmas my true love gave to me: %s in a Pear Tree."
return {
  recite = function(start_verse, end_verse)
    local verses = setmetatable({}, { __index = table })
    for v = start_verse, end_verse do
      verses:insert(verse:format(ordinals[v], table.concat({ table.unpack(subjects, #subjects - v + 1, #subjects) })))
    end
    return verses
  end
}
